### PR TITLE
Provide linear algebra utility

### PR DESCRIPTION
**Pull request by joscao**
Provides basic linear algebra functionality necessary for **array data dependency detection**.

- [x] generate_row_echelon_form:
 
Calculates a row echelon form of a matrix, while allowing a conditional check before each division operation, which can also be specialized, as in our case integers are of interest with proper integer calculation. 
- [x] Is modified version of a stack overflow post, is that a problem?

- [x] back_substitution

Applies a back substitution, provided by an upper triangular matrix onto a right hand side, again optional specialized division operation.    

- [x] is_independent_system 

Checks if a system of equations is independent, i.e. only one coefficient per row for the matrix or none.

- [x] yield_one_d_systems 

Yields the one dimensional systems provided by an independent system of equations, disregards completely zero rows (matrix and rhs). Additionally catches (since it is generalized for inequalities (to be precise <=, >=)  if the lhs row is all zeros and the rhs is non zero, will yield those as well.

~~- [x] bounds_of_one_d_system:~~

~~Calculates all unique bounds (C x > 0) and (Cx < 0) for a one dimensional system.~~
~~- [x] deal properly with zero in coefficient matrix C --> moving bounds to be (C x >= 0 and C x <= 0)~~

### CODE REVIEW

**Comment by joscao** [comment]
Remove unique bounds (C x > 0) and (Cx < 0) for a one dimensional system as it is unclear on how it would deal with zero left hand side , i.e. 0 <= 1 or 0 <= -1 , are those upper bounds, or would they result in failure. Therefore function design is lacking, decided to remove it completly! 
[f982037](https://github.com/ecmwf-ifs/loki/pull/179/commits/f98203716d37a58d373c733c50b97ff942251d43)

File: loki/analyse/util_linear_algebra.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+    "is_independent_system",
+    "yield_one_d_systems",
+]
+
+
+def is_independent_system(matrix):
+    """
+    Check if a linear system of equations can be split into independent one-dimensional problems.
+
+    Args:
```
Comment (reuterbal) [inline loki/analyse/util_linear_algebra.py:24]: Would you mind changing the docstring style to the numpy layout in all routines?
https://www.sphinx-doc.org/en/master/usage/extensions/example_numpy.html

Comment (joscao) [reply loki/analyse/util_linear_algebra.py]: 7de1b4390cc64c2e1895f72356de18b883b4a6f7

File: loki/analyse/util_linear_algebra.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
+# (C) Copyright 2018- ECMWF.
+# This software is licensed under the terms of the Apache Licence Version 2.0
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+# In applying this licence, ECMWF does not waive the privileges and immunities
+# granted to it by virtue of its status as an intergovernmental organisation
+# nor does it submit to any jurisdiction.
+
+from numpy import zeros_like, dot
```
[the remark this answers is not in the document]
Comment (joscao) [reply loki/analyse/util_linear_algebra.py]: 7de1b4390cc64c2e1895f72356de18b883b4a6f7

File: loki/analyse/util_linear_algebra.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+    This function computes the Row Echelon Form (RREF) of a given matrix A using Gaussian elimination.
+    You can provide a custom conditional_check function to add checks or operations during the computation.
+    You can also specify a custom division_operation function for division (e.g., for custom division operations).
+
+    Note:
+    - If the input matrix has no rows or columns, it is already in REF, and the function returns itself.
+    - The function utilizes the specified division operation (default is standard division) for division.
+
+    Reference: https://math.stackexchange.com/questions/3073083/how-to-reduce-matrix-into-row-echelon-form-in-numpy
```
[the remark this answers is not in the document]
Comment (joscao) [reply loki/analyse/util_linear_algebra.py]: 7de1b4390cc64c2e1895f72356de18b883b4a6f7

**Comment by joscao** [comment]
Sorry, discovered a small bug upstream therefore needed to modify this utility. Now fixed. 1754f106bac18801f7aebcfc82a5ad61ba6637fb

**Pull request merged**